content/ui-patterns/saving.mdx

```diff
@@ -0,0 +1,225 @@
+---
+title: Saving
+---
+
+import {Box, Text} from '@primer/components'
+
+<Box sx={{fontSize: 3}} class="lead" as="p">
+    Save patterns help users store and update their content and configuration throughout GitHub. These changes should be represented in the UI accurately, quickly, and obviously. Their behavior should inspire confidence and trust.
+</Box>
+
+## Calls to action
+
+When data is not automatically saved after the user make changes, buttons are used to submit or cancel the changes.
+
+### Text
+
+All configuration flows should have calls to action that include an obvious active verb such as "Create", "Save", "Delete", or "Update" (for example, "Create" a new repository, "Add" an SSH key to your profile, "Save" security preferences, "Update" a repository's description, "Delete" an old email address).
+
+When defining these calls to actions, make sure to:
+1. Keep the text as succinct and clear as possible
+2. Add the item’s name to the button text in cases where it may be unclear to the user what is being changed
+
+### Appearance
+
+If the save button is used to save every input on the page, use the _primary_ button appearance.
+
+If the save button has a corresponding cancel button, use the _primary_ button appearance for the save button, and _secondary_ for the cancel button.
+
+If the save button only refers to a segment of inputs on the page, use the _secondary_ button appearance.
+
+For more guidance on button usage, see the [button interface guidelines](/ui-patterns/button-usage).
+
+### State
+
+Do not disable or hide a form's save button even if the form is invalid or has not been changed. Disabled buttons cannot be focused using the Tab key, and disabled button styles are typically low-contrast and difficult to read.
```

Comment:
When they try and submit, the validation errors would appear and explain why the form couldn't be submitted.

Reply:
My only issue with this is that it seems like better UX to flag invalid input at input time rather than after form submission when the person thinks they're done. Better to prevent than correct when possible.

I also don't see any guidelines around when and how to surface invalid input at field-level, either before or after save. Some examples of that would be useful.

Reply:
@dylanatsmith - validation is a big enough topic that it will have its own interface guidelines. That being said, I can still try and address your feedback for now:

> ...it seems like better UX to flag invalid input at input time rather than after form submission when the person thinks they're done. Better to prevent than correct when possible.

I think there may be some cases where it's good to validate during input (for example: creating a repo name), but I think this pattern usually feels aggressive and distracting.

---

> I also don't see any guidelines around when and how to surface invalid input at field-level

We should surface invalid input at field-level whenever possible. We don't want users to have to guess which fields are invalid and why.

Reply:
Thanks @mperrotti. Sounds like the forthcoming guidelines for validation will clarify those topics.

> I think this pattern usually feels aggressive and distracting.

I get that depending on context, number of fields, requiredness, importance, etc. I don’t feel super strongly about this though.

Reply:
@dylanatsmith I'm happy to chat more about the accessibility implications of inline validation vs. on submit on slack or a zoom call if you'd like! I think form validation is something we'd like to have in our accessibility docs for future reference of deeper questions like this.